src/signature.js

```diff
@@ -50,250 +63,272 @@ const spellNumber = x => ({
+
+  // --------------------------------------------------------------------------
+  //
+  // Types
+  //
+  // --------------------------------------------------------------------------
+
+  // fromUnaryType :: Type -> (Type -> Type)
+  const fromUnaryType = t => $.UnaryType
+    (t.name)
+    (t.url)
+    (t._test)
+    (t.types.$1.extractor);
+
+  // fromBinaryType :: Type -> (Type -> Type -> Type)
+  const fromBinaryType = t => $.BinaryType
+    (t.name)
+    (t.url)
+    (t._test)
+    (t.types.$1.extractor)
+    (t.types.$2.extractor);
+
+  // constructType :: (Array Type) -> Type|Function -> Type
+  const constructType = argTypes => S.ifElse
+    (S.is ($.AnyFunction))
+    (f => f (...argTypes))
+    (t => {
+      assertTypeArity (t, argTypes);
+      switch (t.type) {
+        case 'BINARY':
+          return fromBinaryType (t) (argTypes[0]) (argTypes[1]);
+        case 'UNARY':
+          return fromUnaryType (t) (argTypes[0]);
+        default: {
+          throw new TypeError (`Type ${t.name} should be recreated with `
+            + `Types: ${S.map (S.prop ('name'), argTypes)} but it haven't got `
+            + `a proper function recreator for type ${t.type}.`);
+        }
+      }
+    });
+
+  // Helper Type to wipe out thunks
+  const Thunk = $.NullaryType ('hm-def/Thunk') ('') (S.K (false));
+
+  // convertType :: SignatureEntry -> Reader (TypeMap Type)
+  const convertType = memoize (entry => cond ([
+      [typeEq ('typeConstructor'), convertTypeConstructor],
+      [typeEq ('function'), convertFunction],
+      [typeEq ('list'), convertList],
+      [typeEq ('record'), convertRecord],
+      [typeEq ('constrainedType'), convertConstrainedType],
+      [typeEq ('typevar'), S.pipe ([convertTypevar, Reader.of])],
+      [typeEq ('thunk'), S.K (Reader.of (Thunk))],
+      [S.K (true), e => {
+        throw new Error
+          (`Don't know what to do with signature entry ${e.type}`);
+      }],
+    ]) (entry));
+
+  // convertTypes :: Array SignatureEntry -> Reader (TypeMap (Array Type))
+  const convertTypes = memoize (xs => S.pipe ([
+      S.map (convertType),
+      unchecked.sequence (Reader),
+      unchecked.map (S.reject (S.equals (Thunk))),
+    ]) (xs));
+
+  // convertTypeConstructor :: SignatureEntry -> Reader (TypeMap Type)
+  const convertTypeConstructor = memoize (S.ifElse
+    (hasChildren)
+    (y => S.pipe ([
+      children,
+      convertTypes,
+      x => unchecked.lift2 (constructType) (x) (lookupType (y)),
+    ]) (y))
+    (lookupType));
+
+  // convertList :: SignatureEntry -> Reader (TypeMap Type)
+  const convertList = memoize (S.pipe ([
+    firstChild,
+    convertType,
+    unchecked.map ($.Array),
+  ]));
+
+  // convertFunction :: SignatureEntry -> Reader (TypeMap Type)
+  const convertFunction = memoize (S.pipe ([
+    children,
+    convertTypes,
+    unchecked.map (types => S.reduce
+      (f => x => $.Function ([x, f]))
+      (S.last (types).value)
+      (S.dropLast (1) (types).value)),
```

Comment:
If we can guarantee that `types` is not `[]`, it would be better to use `types[types.length - 1]` and `types.slice (0, -1)`. The `value` property of a Just should be considered private.

Reply:
You are absolutely correct. Don't quite know why I did it this way.